src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractImportCompletionProvider.cs

```diff
@@ -0,0 +1,229 @@
+using Microsoft.CodeAnalysis.AddImports;
+using Microsoft.CodeAnalysis.EditAndContinue;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract class AbstractImportCompletionProvider : CommonCompletionProvider
+    {
+        protected abstract Task<SyntaxContext> CreateContextAsync(Document document, int position, CancellationToken cancellationToken);
+        protected abstract ImmutableArray<string> GetImportedNamespaces(SyntaxNode location, SemanticModel semanticModel, CancellationToken cancellationToken);
+        protected abstract Task<bool> IsInImportsDirectiveAsync(Document document, int position, CancellationToken cancellationToken);
+        protected abstract bool ShouldProvideCompletion(Document document, SyntaxContext syntaxContext);
+        protected abstract Task AddCompletionItemsAsync(CompletionContext completionContext, SyntaxContext syntaxContext, HashSet<string> namespacesInScope, bool isExpandedCompletion, CancellationToken cancellationToken);
+
+        internal override bool IsExpandItemProvider => true;
+
+        private bool? _isImportCompletionExperimentEnabled = null;
+
+        private bool IsExperimentEnabled(Workspace workspace)
+        {
+            if (!_isImportCompletionExperimentEnabled.HasValue)
+            {
+                var experimentationService = workspace.Services.GetService<IExperimentationService>();
+                _isImportCompletionExperimentEnabled = experimentationService.IsExperimentEnabled(WellKnownExperimentNames.TypeImportCompletion);
+            }
+
+            return _isImportCompletionExperimentEnabled == true;
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext completionContext)
+        {
+            var cancellationToken = completionContext.CancellationToken;
+            var document = completionContext.Document;
+
+            // We need to check for context before option values, so we can tell completion service that we are in a context to provide expanded items
+            // even though import completion might be disabled. This would show the expander in completion list which user can then use to explicitly ask for unimported items.
+            var syntaxContext = await CreateContextAsync(document, completionContext.Position, cancellationToken).ConfigureAwait(false);
+            if (!ShouldProvideCompletion(document, syntaxContext))
+            {
+                return;
+            }
+
+            completionContext.ExpandItemsAvailable = true;
```

Comment:
what's this?

Reply:
why do you need to do this if you already exported the member that said you were IsExpandItemProvider ?  I don't get the purpose of having both flags.

Reply:
There are two possible cases if expand provider returns nothing, first is there's actually nothing, another is we might be able to return something but didn't (e.g. timed out). We want to grey out the expander in the first case but show it as selectable in the latter.

Reply:
can you doc this?  i honestly find it very confusing still.  i would prefer ths all be unified somehow .

Reply:
There's already doc for `ExpandItemsAvailable` property